src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -57,11 +61,274 @@ static void RecordDistributedRelationDependencies(Oid distributedRelationId,
 												  Node *distributionKey);
 static ShardPlacement * TupleToShardPlacement(TupleDesc tupleDesc,
 											  HeapTuple heapTuple);
+static List * ShardIntervalsOnWorkerNode(WorkerNode *workerNode, Oid relationId);
+static StringInfo GenerateSizeQueryOnMultiplePlacements(Oid distributedRelationId,
+														List *shardIntervalList,
+														char *sizeQuery);
+static uint64 DistributedTableSizeOnWorker(WorkerNode *workerNode, Oid relationId,
+										   char *sizeQuery);
+static uint64 DistributedTableSize(Oid relationId, char *sizeQuery);
+static void ErrorIfNotSuitableToGetSize(Oid relationId);
 
 
 /* exports for SQL callable functions */
 PG_FUNCTION_INFO_V1(master_stage_shard_row);
 PG_FUNCTION_INFO_V1(master_stage_shard_placement_row);
+PG_FUNCTION_INFO_V1(citus_table_size);
+PG_FUNCTION_INFO_V1(citus_total_relation_size);
+PG_FUNCTION_INFO_V1(citus_relation_size);
+
+
+/*
+ * citus_total_relation_size accepts a table name and
+ * returns a distributed table and its indexes' total relation size.
+ */
+Datum
+citus_total_relation_size(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	uint64 totalRelationSize = 0;
+
+	totalRelationSize = DistributedTableSize(relationId,
+											 INNER_SHARD_TOTAL_RELATION_SIZE_QUERY);
+
+	PG_RETURN_INT64(totalRelationSize);
+}
+
+
+/*
+ * citus_table_size accepts a table name and
+ * returns a distributed table's total relation size.
+ */
+Datum
+citus_table_size(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	uint64 tableSize = 0;
+
+	tableSize = DistributedTableSize(relationId, INNER_SHARD_TABLE_SIZE_QUERY);
+
+	PG_RETURN_INT64(tableSize);
+}
+
+
+/*
+ * citus_relation_size accept a table name and
+ * returns a relation's 'main' fork's size.
+ */
+Datum
+citus_relation_size(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	uint64 relationSize = 0;
+
+	relationSize = DistributedTableSize(relationId, INNER_SHARD_RELATION_SIZE_QUERY);
+
+	PG_RETURN_INT64(relationSize);
+}
+
+
+/*
+ * DistributedTableSize is a helper function for all citus size functions. It
+ * first checks whether the table is distributed and size query can be run on it.
+ * Connection to each node has to be established to get the size of the table.
+ */
+static uint64
+DistributedTableSize(Oid relationId, char *sizeQuery)
+{
+	Relation pgDistNode = NULL;
+	List *workerNodeList = NULL;
+	ListCell *workerNodeCell = NULL;
+	uint64 totalRelationSize = 0;
+
+	if (XactModificationLevel == XACT_MODIFICATION_MULTI_SHARD)
+	{
+		ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+						errmsg("single-shard DML commands must not appear in "
```

Comment:
I think error message should be something like "citus size functions cannot be called in transaction blocks which contain multi-shard data modifications"